### PR TITLE
fixed issue where we crash due to our pending async work run after VS…

### PR DESCRIPTION
**Pull request by heejaechang**
… shutdown

this is another case where we have a pending async task that runs after VS shutdown and it throws and our fail fast code catch that exception and crash VS.

a general fix will be something like us making our fail fast code to aware shutdown situation and ignore any exception if we are in the shutdown situation.

but until we come up with proper design, this should handle one of the high watson hits.

fixes 
https://devdiv.visualstudio.com/DevDiv/_workitems/edit/786353
https://devdiv.visualstudio.com/DevDiv/_workitems/edit/786373

[Watson] clr20r3: CLR_EXCEPTION_System.IO.IOException_8007006d_Microsoft.ServiceHub.Client.dll!Microsoft.ServiceHub.Utility.WrappedStream+_WriteAsync_d__49.MoveNext
[Watson] clr20r3: CLR_EXCEPTION_System.IO.IOException_80131620_Microsoft.ServiceHub.Client.dll!Microsoft.ServiceHub.Utility.WrappedStream+_WriteAsync_d__49.MoveNext

### CODE REVIEW

File: src/VisualStudio/Core/Def/Implementation/Remote/ServiceHubRemoteHostClient.cs [hunk 1]
```diff
@@ -210,12 +218,32 @@ private void UnregisterGlobalOperationNotifications()
...
+            }
+        }
+
         private void OnGlobalOperationStarted(object sender, EventArgs e)
         {
             lock (_globalNotificationsGate)
             {
                 _globalNotificationsTask = _globalNotificationsTask.ContinueWith(
-                    continuation, CancellationToken.None, TaskContinuationOptions.None, TaskScheduler.Default).Unwrap();
+                    continuation, _shutdownCancellationTokenSource.Token, TaskContinuationOptions.None, TaskScheduler.Default).Unwrap();
```
Comment (sharwell) [inline src/VisualStudio/Core/Def/Implementation/Remote/ServiceHubRemoteHostClient.cs]: :question: This omits `LazyCancellation`. Is there a reason why we cannot use that here (e.g. via `SafeContinueWith`)?

File: src/VisualStudio/Core/Def/Implementation/Remote/ServiceHubRemoteHostClient.cs [hunk 2]
```diff
@@ -239,8 +266,7 @@ private void OnGlobalOperationStopped(object sender, GlobalOperationEventArgs e)
             lock (_globalNotificationsGate)
             {
                 _globalNotificationsTask = _globalNotificationsTask.ContinueWith(
-                    continuation, CancellationToken.None,
-                    TaskContinuationOptions.None, TaskScheduler.Default).Unwrap();
+                    continuation, _shutdownCancellationTokenSource.Token, TaskContinuationOptions.None, TaskScheduler.Default).Unwrap();
```
Comment (sharwell) [inline src/VisualStudio/Core/Def/Implementation/Remote/ServiceHubRemoteHostClient.cs]: 📝 Also missing lazy cancellation

File: src/VisualStudio/Core/Def/Implementation/Remote/ServiceHubRemoteHostClient.cs [hunk 3]
```diff
@@ -31,6 +31,7 @@ private enum GlobalNotificationState
 
         private readonly JsonRpc _rpc;
```
Comment (sharwell) [inline src/VisualStudio/Core/Def/Implementation/Remote/ServiceHubRemoteHostClient.cs]: ❓ Why is this not using `JsonRpcEx`, which already handles this case?

File: src/VisualStudio/Core/Def/Implementation/Remote/ServiceHubRemoteHostClient.cs (hunk 1 above)
Comment (heejaechang) [reply src/VisualStudio/Core/Def/Implementation/Remote/ServiceHubRemoteHostClient.cs]: sure.

File: src/VisualStudio/Core/Def/Implementation/Remote/ServiceHubRemoteHostClient.cs (hunk 2 above)
Comment (heejaechang) [reply src/VisualStudio/Core/Def/Implementation/Remote/ServiceHubRemoteHostClient.cs]: sure.

File: src/VisualStudio/Core/Def/Implementation/Remote/ServiceHubRemoteHostClient.cs (hunk 3 above)
Comment (heejaechang) [reply src/VisualStudio/Core/Def/Implementation/Remote/ServiceHubRemoteHostClient.cs]: JsonRpcEx is an abstract type. and it was not designed for this connection.

this connection is the master control connection. but sure I can create a concrete type for this case, but just looking through code, that require some refactoring, so I guess that is out of this PR's scope for now. but sure, I can open an issue for that.

https://github.com/dotnet/roslyn/issues/34555

**Pull request merged**